src/Identity/Core/src/LoggingExtensions.cs

```diff
@@ -0,0 +1,91 @@
+
+using System;
+
+namespace Microsoft.Extensions.Logging
+{ 
+    internal static class LoggingExtensions
+    {
+        private static readonly Action<ILogger, Exception> _invalidExpirationTime;
+        private static readonly Action<ILogger, Exception> _userIdsNotEquals;
+        private static readonly Action<ILogger, Exception> _purposeNotEquals;
+        private static readonly Action<ILogger, Exception> _unexpectedEndOfInput;
+        private static readonly Action<ILogger, Exception> _securityStampNotEquals;
+        private static readonly Action<ILogger, Exception> _securityStampIsNotEmpty;
+        private static readonly Action<ILogger, Exception> _unhandledException;
+
+        static LoggingExtensions()
+        {
+            _invalidExpirationTime = LoggerMessage.Define(
+                eventId: new EventId(0, "InvalidExpirationTime"),
+                logLevel: LogLevel.Information,
+                formatString: "ValidateAsync failed: the expiration time is invalid.");
+
+            _userIdsNotEquals = LoggerMessage.Define(
+                eventId: new EventId(1, "UserIdsNotEquals"),
+                logLevel: LogLevel.Debug,
+                formatString: "ValidateAsync failed: the expected and actual UserId aren't equal");
+
+            _purposeNotEquals = LoggerMessage.Define(
+                eventId: new EventId(2, "PurposeNotEquals"),
+                logLevel: LogLevel.Debug,
+                formatString: "ValidateAsync failed: the expected and actual purpose aren't equal.");
+
+            _unexpectedEndOfInput = LoggerMessage.Define(
+                eventId: new EventId(3, "UnexpectedEndOfInput"),
+                logLevel: LogLevel.Debug,
+                formatString: "ValidateAsync failed: unexpected end of input.");
+
+            _securityStampNotEquals = LoggerMessage.Define(
+                eventId: new EventId(4, "SecurityStampNotEquals"),
+                logLevel: LogLevel.Debug,
+                formatString: "ValidateAsync failed: the expected and actual stamp aren't equal.");
+
+            _securityStampIsNotEmpty = LoggerMessage.Define(
+                eventId: new EventId(5, "SecurityStampIsNotEmpty"),
+                logLevel: LogLevel.Debug,
+                formatString: "ValidateAsync failed: the expected stamp is not empty.");
+
+            _unhandledException = LoggerMessage.Define(
+                eventId: new EventId(6, "UnhandledException"),
+                logLevel: LogLevel.Debug,
+                formatString: "ValidateAsync failed: unhandled exception was thrown.");
+        }
+
+        public static void InvalidExpirationTime(this ILogger logger)
+        {
+            _invalidExpirationTime(logger, null);
+        }
+
+        public static void UserIdsNotEquals(this ILogger logger)
+        {
+            _userIdsNotEquals(logger, null);
+        }
+
+        public static void PurposeNotEquals(this ILogger logger)
```

Comment:
Take in expectedPurpose and actualPurpose and display those to help debugging

Reply:
Ok, I will do it. Thank you for your suggestions